File: seeker/snippet/inheritance.py
#date: 2024-03-27T17:06:11Z
#url: https://api.github.com/gists/cbcab9f24eaa5c169194e4838db3c56f
#owner: https://api.github.com/users/JamesMenetrey

from abc import abstractmethod, abstractproperty, ABC
from typing import List


def load_from_file(filename):
    return [GuineaPig("Kéké"), GuineaPig("Choki"), Cat("Gaspard")]









class Animal(ABC):
    def __init__(self, name):
        self._name = name

    @abstractmethod
    def say_hello(self):
        pass


class Shop:
    def __init__(self, animals: List[Animal]):
        self.__animals = animals

    def hello_from_all(self):
        for animal in self.__animals:
            animal.say_hello()


class GuineaPig(Animal):
    def __init__(self, name):
        super().__init__(name)

    def say_hello(self):
        print(f"{self._name} said: Uiiik")


class Cat(Animal):
    def __init__(self, name):
        super().__init__(name)

    def say_hello(self):
        print(f"{self._name} said: Moew!")


my_shop = Shop(load_from_file("my_shop.txt"))
my_shop.hello_from_all()

File: seeker/snippet/inheritance.java
//date: 2024-07-18T16:45:48Z
//url: https://api.github.com/gists/0ade80bfa276b63466c942e83890acdb
//owner: https://api.github.com/users/docsallover

class Animal {
  String name;
  int age;

  public void eat() {
    System.out.println(name + " is eating.");
  }
}

class Dog extends Animal {
  String breed;

  public void bark() {
    System.out.println(name + " is barking!");
  }
}

public class Main {
  public static void main(String[] args) {
    Dog myDog = new Dog();
    myDog.name = "Buddy";
    myDog.age = 3;
    myDog.breed = "Labrador";

    myDog.eat(); // Inherited from Animal
    myDog.bark(); // Specific to Dog class
  }
}